ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,480 @@
+use crate::isl::isl_type_reference::NullabilityModifier;
+use crate::isl::version_based_isl::constraints::IslConstraint;
+use ion_rs::element::Element;
+use ion_rs::IonType;
+use std::marker::PhantomData;
+
+pub mod constraints;
+
+pub trait IslVersionTrait {}
+
+pub struct IslV1_0 {}
+impl IslVersionTrait for IslV1_0 {}
+
+// TODO: This can be a trait if we have get a new minor version of ISL
+#[derive(Debug, Clone)]
+pub struct IslV2_0 {}
+impl IslVersionTrait for IslV2_0 {}
+
+/// Represents user reserved fields that declares reserved symbols to be used for open content fields.
+#[derive(Debug, Clone, Default, PartialEq)]
+pub struct UserReservedFields {
+    schema_header_fields: Vec<String>,
+    schema_footer_fields: Vec<String>,
+    type_fields: Vec<String>,
+}
+
+impl UserReservedFields {
+    pub fn new(
+        schema_header_fields: Vec<String>,
+        schema_footer_fields: Vec<String>,
+        type_fields: Vec<String>,
+    ) -> Self {
+        Self {
+            schema_header_fields,
+            schema_footer_fields,
+            type_fields,
+        }
+    }
+
+    pub fn schema_header_fields(&self) -> &Vec<String> {
+        &self.schema_header_fields
+    }
+
+    pub fn schema_footer_fields(&self) -> &Vec<String> {
+        &self.schema_footer_fields
+    }
+
+    pub fn type_fields(&self) -> &Vec<String> {
+        &self.type_fields
+    }
+}
+
+/// Represents a schema header
+#[derive(Debug, Clone)]
+pub struct IslSchemaHeader<V: IslVersionTrait> {
+    /// Represents open content as `Element`s in the schema header
+    open_content: Vec<Element>,
+    /// Represents all the IslImports inside the schema file.
+    /// For more information: https://amazon-ion.github.io/ion-schema/docs/isl-1-0/spec#imports
+    imports: Vec<IslImport>,
+    /// Represents the use reserved fields that are defined within schema header,
+    user_reserved_fields: UserReservedFields,
+    _version: PhantomData<V>,
+}
+
+impl<V: IslVersionTrait> IslSchemaHeader<V> {
+    pub fn open_content(&self) -> &[Element] {
+        &self.open_content
+    }
+
+    pub fn imports(&self) -> &[IslImport] {
+        &self.imports
+    }
+}
+
+impl IslSchemaHeader<IslV1_0> {
+    pub fn new(open_content: Vec<Element>, imports: Vec<IslImport>) -> Self {
+        Self {
+            open_content,
+            imports,
+            user_reserved_fields: UserReservedFields::default(),
+            _version: Default::default(),
+        }
+    }
+}
+
+impl IslSchemaHeader<IslV2_0> {
+    pub fn new(open_content: Vec<Element>, imports: Vec<IslImport>) -> Self {
+        Self {
+            open_content,
+            imports,
+            user_reserved_fields: UserReservedFields::default(),
+            _version: Default::default(),
+        }
+    }
+
+    pub fn with_user_reserved_fields(
+        self,
+        schema_header_fields: Vec<String>,
+        schema_footer_fields: Vec<String>,
+        type_fields: Vec<String>,
+    ) -> Self {
```

Comment:
At a glance, I expected this to be two methods (`with_user_reserved_header_fields()` and `with_user_reserved_footer_fields()`). I'd only keep them as a unified method if it's illegal to add one without the other.